client/packages/system/src/Immunisation/VaccineCourseEditModal/VaccineCourseEditModal.tsx

```diff
@@ -0,0 +1,324 @@
+    }
+    updatePatch({ doses: value });
+  };
+
+  // const updateSchedule = (value: number) => {
+  //   if (!value) {
+  //     return;
+  //   }
+  //   const scheduleSeed = (number: number): VaccineCourseScheduleNode => {
+  //     return {
+  //       __typename: 'VaccineCourseScheduleNode',
+  //       id: FnUtils.generateUUID(),
+  //       doseNumber: number,
+  //       label: '',
+  //     };
+  //   };
+  //   let rows = draft?.vaccineCourseSchedules ?? [];
+
+  //   if (rows.length === value) {
+  //     return;
+  //   } else if (value > rows.length) {
+  //     let toAdd = value - rows.length;
+  //     while (toAdd > 0) {
+  //       const number = value - toAdd + 1;
+  //       rows.push(scheduleSeed(number));
+  //       toAdd--;
+  //     }
+  //   } else {
+  //     rows = rows.slice(0, value);
+  //   }
+  //   updatePatch({ vaccineCourseSchedules: rows });
+  // };
+
+  // const updateDescription = (patch: RecordPatch<VaccineCourseScheduleNode>) => {
+  //   if (!patch) {
+  //     return;
+  //   }
+  //   const schedule = {
+  //     ...ArrayUtils.toObject(draft.vaccineCourseSchedules ?? [defaultRow]),
+  //     [patch.id]: patch,
+  //   };
+
+  //   const scheduleArray = Object.values(
+  //     schedule
+  //   ) as VaccineCourseScheduleNode[];
+  //   updatePatch({ vaccineCourseSchedules: scheduleArray });
+  // };
+
+  // const dosesColumns = useColumns<VaccineCourseScheduleNode>(
+  //   [
+  //     { key: 'doseNumber', label: 'label.dose-number' },
+  //     [descriptionColumn(t('label.new-row')), { setter: updateDescription }],
+  //   ],
+  //   {},
+  //   [draft]
+  // );
+
+  const options = useMemo(
+    () => getDemographicOptions(demographicData?.nodes ?? []),
+    [demographicData]
+  );
+
+  if (isLoading) {
+    return <BasicSpinner />;
+  }
+
+  const defaultValue = {
+    value: draft.demographicIndicator?.name ?? '',
+    label: draft.demographicIndicator
+      ? `${draft.demographicIndicator?.name} ${draft.demographicIndicator?.baseYear}`
+      : '',
+  };
+
+  return (
+    <Modal
+      title=""
+      cancelButton={<DialogButton variant="cancel" onClick={onClose} />}
+      okButton={
+        <DialogButton
+          disabled={!isDirty || !programId}
+          variant="ok"
+          onClick={async () => {
+            setIsDirty(false);
+            try {
+              const result =
+                mode === ModalMode.Update
+                  ? await update()
+                  : await create(programId ?? '');
+              if (result?.__typename === 'VaccineCourseNode') {
+                const message =
+                  mode === ModalMode.Update
+                    ? `${t('messages.updated-new-vaccine-course')}: ${
+                        result.name
+                      }`
+                    : `${t('messages.created-new-vaccine-course')}: ${
+                        result.name
+                      }`;
+                success(message)();
+                onClose();
```

Comment:
I preferred to only close the modal if the save or update was successful which would mean proposed changes aren't lost.

Though maybe close and snack indication is how we are operating in the rest of the app?